numpy/core/numeric.py

```diff
@@ -289,16 +288,14 @@ def full(shape, fill_value, dtype=None, order='C'):
     >>> np.full((2, 2), np.inf)
     array([[ inf,  inf],
            [ inf,  inf]])
-    >>> np.full((2, 2), 10, dtype=np.int)
+    >>> np.full((2, 2), 10)
     array([[10, 10],
            [10, 10]])
 
     """
+    if dtype is None:
+        dtype = array(fill_value).dtype
```

Comment:
I wonder about stupid objects, but I guess we just shouldn't care, anyone who wants to do `np.full(3, (1, 2))` should be prepared to give `dtype=object`. (could make it `copy=False` just for the kicks)